python/lsst/pipe/base/graphBuilder.py

```diff
@@ -564,108 +586,80 @@ def fillDatasetRefs(self, registry, collections, run, *, skipExisting=True):
-                # TODO: we could easily support per-DatasetType skipExisting
-                # (it might make sense to put them in originInfo), and I could
-                # imagine that being useful - it's probably required in order
-                # to support writing initOutputs before QuantumGraph
-                # generation.
-                if run is not None:
+        # collection, if there is an output collection.
+        if run is not None:
+            for datasetType, refs in itertools.chain(self.initIntermediates.items(),
+                                                     self.initOutputs.items(),
+                                                     self.intermediates.items(),
+                                                     self.outputs.items()):
+                _LOG.debug("Resolving %d datasets for intermediate and/or output dataset %s.",
+                           len(refs), datasetType.name)
+                for dataId, unresolvedRef in refs.items():
+                    # TODO: we could easily support per-DatasetType skipExisting
+                    # and I could imagine that being useful - it's probably
+                    # required in order to support writing initOutputs before
+                    # QuantumGraph generation.
                     ref = registry.findDataset(datasetType=datasetType, dataId=dataId, collections=run)
-                else:
-                    ref = None
-                if ref is None:
-                    ref = DatasetRef(datasetType, dataId)
-                elif not skipExisting:
-                    raise OutputExistsError(f"Output dataset {datasetType.name} already exists in "
-                                            f"output RUN collection '{run}' with data ID {dataId}.")
-                scaffolding.refs.append(ref)
-        # Prerequisite dataset lookups are deferred until fillQuanta.
-
-    def fillQuanta(self, registry, collections, *, skipExisting=True):
-        """Define quanta for each task by splitting up the datasets associated
-        with each task data ID.
-
-        This method populates `_TaskScaffolding.quanta`.
-
-        Parameters
-        ----------
-        registry : `lsst.daf.butler.Registry`
-            Registry for the data repository; used for all data ID queries.
-        collections : `lsst.daf.butler.CollectionSearch`
-            Object representing the collections to search for input datasets.
-        skipExisting : `bool`, optional
-            If `True` (default), a Quantum is not created if all its outputs
-            already exist.
-        """
+                    if ref is None:
+                        ref = unresolvedRef
+                    elif not skipExisting:
+                        raise OutputExistsError(f"Output dataset {datasetType.name} already exists in "
+                                                f"output RUN collection '{run}' with data ID {dataId}.")
+                    refs[dataId] = ref
```

Comment:
it seems weird to pull out the ref and then reassign it in the case it is not resolved. I think these branches can be re worked a bit.